contracts/exchangeIssuance/ExchangeIssuanceLeveraged.sol

```diff
@@ -0,0 +1,985 @@
+/*
+    Copyright 2021 Index Cooperative
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+    SPDX-License-Identifier: Apache License, Version 2.0
+*/
+pragma solidity 0.6.10;
+pragma experimental ABIEncoderV2;
+
+import { Address } from "@openzeppelin/contracts/utils/Address.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { IUniswapV2Factory } from "@uniswap/v2-core/contracts/interfaces/IUniswapV2Factory.sol";
+import { IUniswapV2Router02 } from "@uniswap/v2-periphery/contracts/interfaces/IUniswapV2Router02.sol";
+import { Math } from "@openzeppelin/contracts/math/Math.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import { SafeMath } from "@openzeppelin/contracts/math/SafeMath.sol";
+import { ReentrancyGuard } from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+
+import { IAToken } from "../interfaces/IAToken.sol";
+import { IDebtIssuanceModule } from "../interfaces/IDebtIssuanceModule.sol";
+import { IController } from "../interfaces/IController.sol";
+import { ISetToken } from "../interfaces/ISetToken.sol";
+import { IWETH } from "../interfaces/IWETH.sol";
+import { PreciseUnitMath } from "../lib/PreciseUnitMath.sol";
+import { UniSushiV2Library } from "../../external/contracts/UniSushiV2Library.sol";
+import { FlashLoanReceiverBaseV2 } from "../../external/contracts/aaveV2/FlashLoanReceiverBaseV2.sol";
+
+
+
+
+/**
+ * @title ExchangeIssuance
+ * @author Index Coop
+ *
+ * Contract for issuing and redeeming a leveraged Set Token
+ * Supports all tokens with one collateral Position in the form of an AToken and one debt position
+ * Both the underlying of the collateral as well as the debt token have to be available for flashloand and be 
+ * tradeable against each other on Sushi / Quickswap
+ *
+ */
+contract ExchangeIssuanceLeveraged is ReentrancyGuard, FlashLoanReceiverBaseV2 {
+
+    using Address for address payable;
+    using SafeMath for uint256;
+    using PreciseUnitMath for uint256;
+    using SafeERC20 for IERC20;
+    using SafeERC20 for ISetToken;
+
+    /* ============ Enums ============ */
+
+    enum Exchange { None, Quickswap, Sushiswap}
```

Comment:
Good point. Will do this in a separate if we do want to add it. 

On this occasion I'd also like to reraise the question of wether / how we might want to factor out this whole "exchange adapter" logic into a library or a separate contract. (Beside improving readability and reusability it might also allow us to later update the "exchange adapter" logic for example adding new exchanges etc.)

Reply:
I'm thinking about extracting the "exchange adapter" logic into a separate contract from which to inherit, to improve readability / usability.

That way it still won't be upgradeable which could easily changed but I'm not sure what our policy on  upgradability is in general.

Reply:
Yeah I think this is a good idea. Go for it

Reply:
So should we make  the exchange adapter logic upgradeable or just use "normal" inheritance ? (having to redeploy the whole EI contract if we want to update the exchange interface)

Reply:
I think normal inheritance is sufficient here. This contract will be on L2 so redeployment is cheap. It also lets us avoid any added complexity